text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,321 @@
+  addToHasMany(modelName: string, id?: string, clientId?: string, key: string, jsonApiResources, idx: number) {}
+  removeFromHasMany(modelName: string, id?: string, clientId?: string, key: string, jsonApiResources) {}
+}
+```
+
+The above is just one snapshot of the roughly dozen APIs relating to mutation of the ~25 APIs present on `RecordData`. While these methods
+ serve an important role helping to ensure that all of the features of `ember-data`'s cache, relationship layer, and model layer are available
+ to `RecordData`, the "soup" prevents someone from knowing at a glance which methods are for mutations, and whether these methods are for mutating local state, or communicating remote state changes. It also means that developers must memorize a large API surface area, and don't
+ have a clear picture of how these methods relate to each other in terms of what they achieve.
+
+Contrast that with the following (theoretical) API.
+
+```typescript
+interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
+}
+
+interface Operation {
+  op: string;
+}
+
+interface RecordData {
+  performMutation(operation: Operation): void
+}
+```
+
+We could then describe our original `RecordData` methods above as potential mutation `operation`s.
+For example:
+
+```typescript
+
+interface ReplaceAttributeOperation extends Operation {
+  op: 'replaceAttribute';
+  record: RecordIdentifier;
+  attribute: string;
+  value: any;
+}
+interface AddToRelatedRecordsOperation extends Operation {
+  op: 'addToRelatedRecords';
+  record: RecordIdentifier;
+  relationship: string;
+  relatedRecord: RecordIdentifier;
+}
+interface RemoveFromRelatedRecordsOperation extends Operation {
+  op: 'removeFromRelatedRecords';
+  record: RecordIdentifier;
+  relationship: string;
+  relatedRecord: RecordIdentifier;
+}
+```
+
+The benefit of this structure is that not only does it simplify the cognitive overhead of remembering all of the APIs, it unifies all of our
+ mutations into a single mental model that provides an extension point should we desire to add additional features to `ember-data` such as
+ a log of mutations, reverse operations, or transactions (groups of operations that must succeed or fail together).
+
+### Nested Saves / API Transactions
+
+// TODO 
+
+### Websocket Support
+
+// TODO 
+
+### Better Cache Serialization
+
+// TODO 
+
+### Improved Infra for Offline Support
+
+// TODO 
+
+### Improved Infra for Cache and Relationship Management
+
+// TODO 
+
+----
+
+## Detailed design
+
+### New Typings
+
+```typescript
+import { Value as JSONValue } from 'json-typescript';
+
+export type Meta = { [k: string]: JSONValue };
+
+export interface ResourceIdentifier {
+  lid?: string;
+  id: string | null;
+  type: string;
+  meta?: Meta;
+}
+
+export interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
+}
+```
```

Comment:
`ResourceIdentifier` and `RecordIdentifier` have extremely similar names that do not make clear their semantic differences (ie that a record identifier is referentially stable and a resource identifier may not be).

What do you think about renaming `RecordIdentifier` to `StableIdentifier`.

Reply:
Stability is not the difference between the two.

A `ResourceIdentifier` is the `json-api` interface, which may or may not have `lid` and `meta` but MUST have `type` (and must have `id` if not for a serialized version of a new client created resource with an `lid`)
A `RecordIdentifier` is our own interface for internal use, which additionally MUST have `lid` and which always has `meta` (although it may be `null`).

Reply:
cc @dgeb for thoughts, as we are paralleling a similar conceptual parity from orbit here.

Reply:
> @runspired @dgeb can we not make that json-api specific package as part of this change? presumably with the intent to eventually (or perhaps immediately) move it into the json-api org?

Not sure why I can't reply to this comment on the comment, but the TL;DR is that it is unlikely we ever perfectly match the signature of jsonapi here for `RecordIdentifier`, but we may be able to match the official interface for `ResourceIdentifier` once the spec includes the `lid` addition.

Reply:
For the separate package I was referring specifically to `ResourceIdentifier`.  I agree `RecordIdentifier` should be internal.

---

> Stability is not the difference between the two.

> Additionally, our identifiers used internally or given from ember-data to application code will be referentially stable, allowing them to be used as keys in either WeakMap or Map based caches.

Are `RecordIdentifier`s referentially stable?